internal/engine/evaluator.go

```diff
@@ -389,22 +391,22 @@ func (ep evalParams) evaluateBoolCELExpr(expr *exprpb.CheckedExpr, variables map
 	return boolVal, nil
 }
 
-func (ep evalParams) evaluateStrCELExpr(expr *exprpb.CheckedExpr, variables map[string]any, input *enginev1.CheckInput) (string, error) {
+func (ep evalParams) evaluateProtobufValueCELExpr(expr *exprpb.CheckedExpr, variables map[string]any, input *enginev1.CheckInput) (*structpb.Value, error) {
 	val, err := ep.evaluateCELExpr(expr, variables, input)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 
 	if val == nil {
-		return "", nil
+		return nil, nil
 	}
 
-	strVal, ok := val.(string)
-	if !ok {
-		return "", nil
+	pbVal, err := structpb.NewValue(val)
```

Comment:
Unfortunately, this doesn't work out of the box for anything but simple values. You'd need something like https://github.com/cerbos/cerbos/blob/b6ee7591eed7bbb7f363af9d0be7710ff5487f79/internal/engine/tracer/context.go#L173 to convert Go values to `structpb.Value`s. Try with some expressions like `["foo", "bar"]` or `{"id": P.id, "keys": [R.attr.id]}`. 

Also, if evaluating the output expression throws an error, we return DENY but I am not sure whether that's the right thing to do because it's just an optional thing attached to the access rule being enforced. We should probably return a special error value or `null` for the output value but not actually fail the check itself. We should also  log the fact that the expression failed.

Reply:
Probably worth adding some complicated `output` expressions (they don't have to make sense) to some of the policies we have in `internal/test/testdata/store` so that we can catch any regressions or odd behaviours during test runs. It would help to measure the impact of output values in the engine benchmarks as well.

Reply:
I opted to handle output evaluation errors by passing up protobuf strings to feed back to the user (rather than bubbling up the error and tracing the failure). It made sense to me to give immediate feedback to the user, but I'm happy to refactor to pass up and log errors if preferable. Might be better from a performance standpoint, to avoid rendering error strings in the output 🤷‍♂️ .

I'm a little wary of performance penalties in general (as Charithe mentioned), so worth keeping an eye on - particularly for more complex expressions, given that we support arbitrary depth.